src/docs/sphinx/dev_guide/weak_form.rst

```diff
@@ -0,0 +1,261 @@
+     auto f0 = b(x); // evaluate the body-force at the location of the quadrature point
+     auto f1 = zero{}; // for this term, the f_1 term is identically zero
+     
+     return std::tuple{f0, f1};
+     
+   }, mesh);
+
+And finally, for the surface tractions:
+
+.. code-block:: cpp
+
+   // WeakForm::AddSurfaceIntegral() only expects us to return f, so we don't need a tuple
+   residual.AddSurfaceIntegral([](auto x, auto disp /* unused */){
+     return traction(x); // evaluate the traction at the location of the quadrature point
+   }, surface_mesh);
+
+Now that we've finished describing all the integral terms that appear in
+our residual, we can carry out the actual calculation by calling
+``WeakForm::operator()``:
+
+.. code-block:: cpp
+
+   auto r = residual(displacements);
+
+Putting these snippets together without the verbose comments, we have (note: the two AddVolumeIntegrals were fused into one):
+
+.. code-block:: cpp
+
+   using test = H1<order, dim>;
+   using trial = H1<order, dim>;
+   WeakForm< test(trial) > residual(test_fes, trial_fes);
+
+   // note: the first two AddVolumeIntegral calls can be fused
+   // into one, provided they share the same domain of integration
+   residual.AddVolumeIntegral([](auto x, auto disp){
+     return std::tuple{b(x), stress(std::get<1>(disp))};
+   }, mesh);
+
+   residual.AddSurfaceIntegral([](auto x, auto disp /* unused */){ return traction(x); }, surface_mesh);
+
+   auto r = residual(displacements);
+
+So, in only a few lines of code, we can create optimized, custom finite
+element kernels!
+
+.. _header-n34:
+
+Implementation
+==============
+
+For the most part, the ``WeakForm`` class is just a container of
+``Integral`` objects, and some prolongation and restriction operators to
+get the data they need:
+
+.. code-block:: cpp
+
+   template <typename test, typename trial>
+   struct WeakForm<test(trial)> : public mfem::Operator {
+     ...
+     std::vector< Integral<test(trial)> > domain_integrals;
+     std::vector< Integral<test(trial)> > boundary_integrals;
+   };
+
+The calls to ``WeakForm::Add****Integral`` forward the integrand and
+mesh information to an ``Integral`` constructor and add it to the
+appropriate list (either ``domain_integrals`` or
+``boundary_integrals``). MFEM treats domain and boundary integrals
+differently, so we maintain them in separate lists.
+
+From there, the ``Integral`` constructor uses the integrand functor to
+specialize a highly templated finite element kernel (simplified
+implementation given below).
+
+.. code-block:: cpp
+
+   template < ::Geometry g, typename test, typename trial, int geometry_dim, int spatial_dim, int Q,
+              typename derivatives_type, typename lambda>
+   void evaluation_kernel(const mfem::Vector& U, mfem::Vector& R, derivatives_type* derivatives_ptr,
+                          const mfem::Vector& J_, const mfem::Vector& X_, int num_elements, lambda qf)
+   {
```

Comment:
This documentation is great!  Haven't had the chance to read through it completely but would it be possible to use Sphinx `literalinclude`s for these snippets so they stay up-to-date with the source code?

e.g., https://github.com/LLNL/serac/blame/develop/src/docs/sphinx/user_guide/simple_conduction_tutorial.rst#L44-L47